rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,436 @@
+```
+
+with 
+- $A_i$ a set of matrices of dimension $M \times K$
+- $B_i$ a set of matrices of dimension $K \times N$
+- D and C matrices of dimension $M \times N$
+- bias a vector of dimension $N$.
+
+This proposal discusses exposing these sequential, basic building
+blocks externally for end-users to conveniently be able to write their
+own optimized custom primitives/routines. These can be used in their
+own applications and/or contributed back behind the oneDNN graph API.
+
+As a starter, we will expose only simple transforms and brgemm
+routines.
+
+## Memory descriptors, memory objects and/or pointers ?
+
+In general, we need to express memory shapes and memory objects, and
+for this block level API, we will have are three considerations:
+- allowing arbitrary strides between blocks ($A_i$, $B_i$ in
+  particular) for maximum flexibility,
+- Support for metadata (e.g. zero-point handling) and
+  hardware-specific layouts,
+- runtime overheads.
+
+### Arbitrary strides between A and B blocks
+
+To make the API flexible, we want to allow arbitrary strides between A
+and B blocks as showed in the picture:
+![](brgemm_pic.png)
+
+This is necessary on multiple occasions, a few examples being:
+- for convolutions, where blocks can overlap depending on covolution strides
+- for LSTM/GRU where weights for each gate can live in separate memory addresses.
+- for models with embedding lookup, where vectors can be passed as separate blocks.
+
+Using arbitrary strides between blocks is unfortunately not possible
+if we encapsulate all $A_i$ blocks (resp $B_i$ blocks) of a brgemm
+operation in a dnnl::memory object, since this will force a fixed
+stride between each block.
+
+Another option would be to take an array of dnnl::memory objects, each
+containing a single block of A/B. This would require the user to call
+`dnnl::memory::set_data_handle()` on each of these objects before
+brgemm call. It would incur some unnecessary overhead, but should
+not be anything substantial.
+
+However, one might wonder why use memory descriptors at all and not
+just take pointers directly as input to brgemm::execute
+call. Internally we support multiple flavors:
+- array of pointers: here the user will have to do pointer arithmetic
+  before passing those pointers to brgemm::execute call.
+- base pointer + arrray of offsets: here the pointer arithmetic will
+  be handled internally behind brgemm::execute call.
+- base pointer + fixed stride: this flavor has limited application and
+  is typically used for matmul implementation. Given its limited
+  applicability, it seems unnecessary to expose it.
+  
+To make the public brgemm API simpler, we propose to expose only the array
+of pointers flavor.
+
+For the brgemm API specifically, an array of equal sizes for A and B
+blocks are passed.  To guarantee these are of equal size, and simplify
+pointer arithmetic logic inside generated kernels, we will take an
+array of pairs of pointers.
```

Comment:
I am not sure I understand. Could you give an example or more detail on the concern?

Because of the batch reduction semantic, we have to take as many A_i blocks as B_i blocks. Here the two options are:
- take two arrays of equal size (one for A_i, one for B_i)
- take a single array of pairs (to interleave pointers to A_i/B_i)

Reply:
I also don't understand why merging A and B into single array can guarantee they are of equal size. What if the array size is odd? The sizes of A and B pointer array are specified by `batch` parameter. User should guarantee both A and B array contains `batch` number of pointers.

Reply:
@mgouicem, FWIW, taking two vectors will save us from having to copy pointers from the pairs to another C-like array to pass them forward to the C API, which means less overhead.

Reply:
> I also don't understand why merging A and B into single array can guarantee they are of equal size. What if the array size is odd? The sizes of A and B pointer array are specified by `batch` parameter. User should guarantee both A and B array contains `batch` number of pointers.

With the C++ API, because we take an array of pair, the number of pointers in the internal buffer is always even.
But in general, I agree with you, because the API is low level, we will likely do minimal checks on arguments.

Reply:
> C++ to C API. Whether we use 2 vectors or vector of pairs, we can pass those directly. Albeit, it is cleaner with 2 separate vectors, it is still doable without a copy for vector of pairs (see [here](https://godbolt.org/z/o5PT8Wx9c))

@mgouicem, I think there are two problem with your example:
* You are assuming that the data members of `std::pair` are contiguous in memory. The standard doesn't seem to guarantee that. Your example works because of the way `std::pair` is implemented ([link](https://github.com/gcc-mirror/gcc/blob/18e94e04dae724c61cbc13ace85fa68f2deda900/libstdc%2B%2B-v3/include/bits/stl_pair.h#L284)). In our case `std::pair` is just a struct with two pointers therefore no padding is required hence the example works.
* You are basically passing a C++ object (`std::pair`) as `void *` to the C API. By design, we forbid C++ objects from crossing the library boundary (a.k.a C API) unless there is no way around it (e.g. SYCL interop API)

Reply:
I agree standard does not guarantee std::pair is contiguous in memory. However, all compilers that I know of implement it this way. Same as type punning, or reinterpret_cast: wrt to standard, they are not safe to use, but in practice, compilers tend to preserve behaviors that are not required by standard.

In any case, we can replace std::pair with a dedicated struct, no big deal.

Reply:
> In any case, we can replace std::pair with a dedicated struct, no big deal.

If we are dead set on passing pairs of pointers then I think that's what we need to do, this way we won't rely on the implementation details of `std::pair`.

Reply:
So, is the decision here to go with an array of pointer pairs, accepting potentially higher overhead?

Reply:
Also, as a follow-up to the offline conversation, expanding a little on the original comment:

> Can this be a potential problem in the case of some lookup pointer arrays to both A and B? I'm imagining some cases of embeddings. In that case, the API that accepts a vector of pairs would likely force some memory movement.

Consider an example when one of the inputs is a weight matrix and another one is accessed indirectly as you'd do in the case of an embedding lookup (there are also other similar cases found in beam search in LLMs, image attention masks, roi_heads in detection). Restricting the API to accept pairs of pointers hinders optimization possibilities. So, those cases can be addressed with a separate interface, yet the feasibility should also be estimated.